packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -2414,55 +2418,47 @@ final class _RenderDeferredLayoutBox extends RenderProxyBox
   @override
   RenderObject? get debugLayoutParent => _layoutSurrogate;
 
-  void layoutByLayoutSurrogate() {
-    assert(!_theaterDoingThisLayout);
-    final _RenderTheater? theater = parent as _RenderTheater?;
-    if (theater == null || !attached) {
-      assert(false, '$this is not attached to parent');
-      return;
-    }
-    if (theater._layingOutSizeDeterminingChild) {
-      theater.invokeLayoutCallback((BoxConstraints constraints) {
-        markNeedsLayout();
-      });
-    } else {
-      final BoxConstraints theaterConstraints = theater.constraints;
-      final Size boxSize =
-          theaterConstraints.biggest.isFinite
-              ? theaterConstraints.biggest
-              // Accessing the theater's size is only unsafe if it is laying out the
-              // size-determining child.
-              : theater.size;
-      super.layout(BoxConstraints.tight(boxSize));
-    }
-  }
-
-  bool _theaterDoingThisLayout = false;
-  @override
-  void layout(Constraints constraints, {bool parentUsesSize = false}) {
-    assert(_needsLayout == debugNeedsLayout);
-    // Only _RenderTheater calls this implementation.
-    assert(parent != null);
-    final bool scheduleDeferredLayout = _needsLayout || this.constraints != constraints;
-    assert(!_theaterDoingThisLayout);
-    _theaterDoingThisLayout = true;
-    super.layout(constraints, parentUsesSize: parentUsesSize);
-    assert(_theaterDoingThisLayout);
-    _theaterDoingThisLayout = false;
+  /// Whether this RenderBox's layout method is currently being called by the
+  /// theater or the layoutSurrogate's [performLayout] implementation.
+  bool _doingLayoutFromTreeWalk = false;
+  void _doLayoutFrom(RenderObject treewalkParent, {required Constraints constraints}) {
+    final bool shouldAddToDirtyList = needsLayout || this.constraints != constraints;
+    assert(!_doingLayoutFromTreeWalk);
+    _doingLayoutFromTreeWalk = true;
+    super.layout(constraints);
+    assert(_doingLayoutFromTreeWalk);
+    _doingLayoutFromTreeWalk = false;
     _needsLayout = false;
     assert(!debugNeedsLayout);
-    if (scheduleDeferredLayout) {
-      final _RenderTheater parent = this.parent! as _RenderTheater;
+    if (shouldAddToDirtyList) {
+      // Instead of laying out this subtree via treewalk, adding it to the dirty
+      // list. This ensures:
+      //
+      //  1. this node will be laid out by the PipelineOwner *after* the two
```

Comment:
For my own understanding (and worthwhile to document here): the *after* is guaranteed because this RO has a higher depth than the theater and surrogate?